### PR TITLE
feat: add clap and server and client options

### LINKED ISSUE #17: Allow for changing the listening address of the server

### LINKED ISSUE #759: Allow for changing the listening address of the server

### PR DESCRIPTION
**Pull request by ramfox**
- help & version commands
- optional custom port for server
- optional custom server addr for client
- server takes a list of filepaths
- client saves the transfered file to the filesystem. By default the name of this file is the hash. Optional flag for setting your own `out` path

closes n0-computer/iroh#759 n0-computer/iroh#760 n0-computer/iroh#772 n0-computer/iroh#774

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,14 +1,56 @@
...
+
+#[derive(Subcommand, Debug, Clone)]
+enum Commands {
+    /// Serve the data from the given path
+    #[clap(about = "Serve the data from the given path")]
+    Server {
+        paths: Vec<PathBuf>,
+        #[clap(long, short)]
+        /// Optional port
+        /// Defaults to 4433
```
Comment (flub) [inline src/main.rs]: clap merges this on one line and that looks a bit odd.  Perhaps:
```suggestion
        /// Optional port, defaults to 4433
```

Same for the other help strings.

File: src/main.rs [hunk 2]
```diff
@@ -303,15 +388,25 @@ async fn main() -> Result<()> {
 #[cfg(test)]
 mod tests {
     use super::*;
+    use testdir::testdir;
```
Comment (flub) [inline src/main.rs:378]: :tada:

File: Cargo.toml [hunk 3]
```diff
@@ -18,3 +19,6 @@ serde = { version = "1.0.152", features = ["derive"] }
 tokio = { version = "1.24.1", features = ["full"] }
 tokio-util = { version = "0.7", features = ["io-util", "io"] }
 unsigned-varint = { version = "0.7.1", features = ["futures"] }
+
+[dev-dependencies]
+testdir = "0.7.0"
```
Comment (flub) [inline Cargo.toml]: I fixed a minor annoyance, can you upgrade to `0.7.1`?

**Review by flub — APPROVED** [review]
There are compiler warnings still.

File: src/main.rs [hunk 4]
```diff
@@ -130,6 +188,13 @@ async fn main_client(hash: bao::Hash) -> Result<()> {
                     while decoder.read(&mut buf)? > 0 {}
                 }
 
+                let outpath = if let Some(out) = opts.out {
+                    out
+                } else {
+                    // default to name as hash
+                    std::path::PathBuf::from(hash.to_string())
+                };
+                std::fs::write(outpath, in_buffer)?;
```
Comment (dignifiedquire) [inline src/main.rs]: this is an `async` function, should use `tokio::fs`

File: src/main.rs [hunk 5]
```diff
@@ -255,12 +340,16 @@ struct Data {
...
     println!("Available Data:");
 
     let mut db = HashMap::new();
-    for num in range {
-        let data = vec![1u8; 1024 * 1024 * num];
+    for path in paths {
+        if !path.is_file() {
+            panic!("can only transfer blob data");
+        }
+        let data = std::fs::read(path).unwrap();
```
Comment (dignifiedquire) [inline src/main.rs]: Should be an `async` function and use `tokio::fs` now, to not block the caller

File: src/main.rs [hunk 6]
```diff
@@ -303,15 +388,25 @@ async fn main() -> Result<()> {
...
 mod tests {
     use super::*;
+    use testdir::testdir;
 
     #[tokio::test]
     async fn basics() {
-        let db = create_db(vec![1]);
+        let dir: PathBuf = testdir!();
+        let path = dir.join("hello_world");
+        std::fs::write(&path, "hello world!").ok();
```
Comment (dignifiedquire) [inline src/main.rs]: async

File: src/main.rs [hunk 7]
```diff
@@ -303,15 +388,25 @@ async fn main() -> Result<()> {
...
             main_server(db).await.unwrap();
         });
 
-        main_client(hash).await.unwrap();
+        let out = dir.join("out");
+        let mut opts = ClientOptions::default();
+        opts.out = Some(out.clone());
+        main_client_with_options(hash, opts).await.unwrap();
+        let got = std::fs::read(out).ok();
+        let expect = std::fs::read(path).ok();
```
Comment (dignifiedquire) [inline src/main.rs]: async

File: src/main.rs [hunk 8]
```diff
@@ -303,15 +388,25 @@ async fn main() -> Result<()> {
...
         tokio::task::spawn(async move {
             main_server(db).await.unwrap();
         });
 
-        main_client(hash).await.unwrap();
+        let out = dir.join("out");
+        let mut opts = ClientOptions::default();
+        opts.out = Some(out.clone());
+        main_client_with_options(hash, opts).await.unwrap();
+        let got = std::fs::read(out).ok();
```
Comment (dignifiedquire) [inline src/main.rs]: async

File: src/main.rs [hunk 9]
```diff
@@ -255,12 +340,16 @@ struct Data {
...
 
-fn create_db(range: Vec<usize>) -> Arc<HashMap<bao::Hash, Data>> {
+fn create_db(paths: Vec<&Path>) -> Arc<HashMap<bao::Hash, Data>> {
     println!("Available Data:");
 
     let mut db = HashMap::new();
-    for num in range {
-        let data = vec![1u8; 1024 * 1024 * num];
+    for path in paths {
+        if !path.is_file() {
```
Comment (rklaehn) [inline src/main.rs]: anyhow::ensure!(path.is_file(), "can only transfer blob data")

**Comment by dignifiedquire** [comment]
There is a binary file checked in, which I suspect you didn't want

**Comment by dignifiedquire** [comment]
@flub @rklaehn pushed the fixes, as I am not sure if @ramfox is around, so we can get this merged

**Pull request merged**